src/irr/video/COpenGLSpecializedShader.cpp

```diff
@@ -114,10 +114,104 @@ COpenGLSpecializedShader::COpenGLSpecializedShader(const video::IVideoDriver* _d
 
     GLchar logbuf[1u<<12]; //4k
     driver->extGlGetProgramInfoLog(m_GLname, sizeof(logbuf), nullptr, logbuf);
-    os::Printer::log(logbuf, ELL_ERROR);
+    if (logbuf[0])
+        os::Printer::log(logbuf, ELL_ERROR);
 
-    // TODO:
-    // what should be interface for setting push_constants (regular uniform on GL backend) for now?
+    m_introspectionData = core::smart_refctd_ptr<asset::CIntrospectionData>(_introspection);
+}
+
+void COpenGLSpecializedShader::setUniformsImitatingPushConstants(const uint8_t* _pcData)
+{
+    if (m_uniformsList.empty())
+        buildUniformsList();
+
+    using gl = COpenGLExtensionHandler;
+    for (const SUniform& u : m_uniformsList)
+    {
+        const SMember& m = u.m;
+        auto is_mtx = [&m] { return (m.mtxRowCnt>1u && m.mtxColCnt>1u); };
+        auto is_single_or_vec = [&m] { return (m.mtxRowCnt>=1u && m.mtxColCnt==1u); };
+
+        if (is_mtx() && m.type==asset::EGVT_F32) {
+            core::vector<GLfloat> matrix_data(m.mtxRowCnt*m.mtxColCnt*m.count);
+            for (uint32_t i = 0u; i < m.count; ++i)
+            {
+                for (uint32_t c = 0u; c < m.mtxColCnt; ++c) {
+                    const GLfloat* col = reinterpret_cast<const GLfloat*>(_pcData + m.offset + i*m.arrayStride + c*m.mtxStride);
+                    matrix_data.insert(matrix_data.cend(), col, col+m.mtxRowCnt);
+                }
+            }
+            PFNGLPROGRAMUNIFORMMATRIX4FVPROC glProgramUniformMatrixNxMfv_fptr[3][3]{ //N - num of columns, M - num of rows because of weird OpenGL naming convention
+                {&gl::extGlProgramUniformMatrix2fv, &gl::extGlProgramUniformMatrix2x3fv, &gl::extGlProgramUniformMatrix2x4fv},//2xM
+                {&gl::extGlProgramUniformMatrix3x2fv, &gl::extGlProgramUniformMatrix3fv, &gl::extGlProgramUniformMatrix3x4fv},//3xM
+                {&gl::extGlProgramUniformMatrix4x2fv, &gl::extGlProgramUniformMatrix4x3fv, &gl::extGlProgramUniformMatrix4fv} //4xM
+            };
+            glProgramUniformMatrixNxMfv_fptr[m.mtxColCnt-2u][m.mtxRowCnt-2u](m_GLname, u.location, m.count, GL_FALSE, matrix_data.data());
```

Comment:
actually you need to introspect if the layout qualifier `row_major` applies to that particular push_constant

Reply:
yea, I thought about that, but left it for tomorrow. Actually there are spv decorations for it, it's nice. I'll add to introspection data

Reply:
+1

Reply:
resolved however when i looked in spirv-cross code i couldnt see actually gathering this decoration from spv, needs testing